### PR TITLE
feat: include workspace and project as parameters for listing workflows on CE

### PR DESCRIPTION
**Pull request by BenjaminMummery**
# The problem

As an user, I would like to retrieve a list of workflow runs, viewing only runs in a workspace ([ORQSDK-806](https://zapatacomputing.atlassian.net/browse/ORQSDK-806)):

```python
from orquestra import sdk

wf_run_list = sdk.list_workflow_runs(config=my_config, workspace=my_workspace, ...)
```

or viewing only runs in a project ([ORQSDK-830](https://zapatacomputing.atlassian.net/browse/ORQSDK-830)):

```python
from orquestra import sdk

wf_run_list = sdk.list_workflow_runs(config=my_config, workspace=my_workspace, project=my_project...)
```
I also want to be able to do this from the CLI ([ORQSDK-789](https://zapatacomputing.atlassian.net/browse/ORQSDK-789))

# This PR's solution

The behaviour of list workflows differs from the other workspace-related commands because the 'project' parameter will be optional in the platform API.

- Adds workspace and project as optional arguments to the runtime `list_workflow_runs` functions.
  - Non-CE runtimes will raise a WorkspacesNotSupportedError if either of these arguments are actually used.
- Adds `get_space_ids()` to extract a unique workspace and project ID from combinations of workspace and project arguments. It's possible to pass a ProjectRef that contains both to `list_workflow_runs`, so we need to be able to handle a) cases where one argument contains both IDs, b) cases where each ID comes from a separate argument, and c) cases where we have a workspace and no project, as well as guarding against d) having a projectID and no workspace, and e) having a project argument that contains a workspace ID that differs from an ID in the workspace argument.
- Adds workspaceId and projectId to the platform API call.
- Modifies the CLI command `orq wf list` to make filtering by workspace possible. Now accepts workspace as a separate argument and, when prompting for project, includes an `all` option:
    ```
    > orq wf list -w my_workspace
    [?] Project:
    > my_project
       my_other_project
       all
    ```

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`
- [x] The PR is linked to a JIRA ticket (if there's no suitable ticket, check the box).

[ORQSDK-806](https://zapatacomputing.atlassian.net/browse/ORQSDK-806)
[ORQSDK-830](https://zapatacomputing.atlassian.net/browse/ORQSDK-830)
[ORQSDK-789](https://zapatacomputing.atlassian.net/browse/ORQSDK-789)

[ORQSDK-806]: https://zapatacomputing.atlassian.net/browse/ORQSDK-806?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[ORQSDK-830]: https://zapatacomputing.atlassian.net/browse/ORQSDK-830?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ORQSDK-789]: https://zapatacomputing.atlassian.net/browse/ORQSDK-789?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ORQSDK-830]: https://zapatacomputing.atlassian.net/browse/ORQSDK-830?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ORQSDK-789]: https://zapatacomputing.atlassian.net/browse/ORQSDK-789?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by alexjuda — CHANGES_REQUESTED** [review]
As mentioned on Slack, I'm posting my comments for files I already went through

File: src/orquestra/sdk/_base/_api/_wf_run.py [hunk 1]
```diff
@@ -471,9 +472,11 @@ def list_workflow_runs(
     max_age: t.Optional[str] = None,
     state: t.Optional[t.Union[State, t.List[State]]] = None,
     project_dir: t.Optional[t.Union[Path, str]] = None,
-    project: t.Optional[ProjectRef] = None,
+    workspace: t.Optional[t.Union[Workspace, WorkspaceId]] = None,
+    project: t.Optional[t.Union[Project, ProjectRef, ProjectId]] = None,
```
[the remark this answers is not in the document]
Comment (alexjuda) [reply src/orquestra/sdk/_base/_api/_wf_run.py:476]: [Discussed elsewhere]: let's go with alternative (1)

File: src/orquestra/sdk/_base/_spaces/_resolvers.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import typing as t
+
+from orquestra.sdk.schema.workflow_run import ProjectId, WorkspaceId
+
+from .._spaces._structs import Project, ProjectRef, Workspace
+
+
+def get_space_ids(
+    workspace: t.Optional[t.Union[Workspace, WorkspaceId]],
+    project: t.Optional[t.Union[Project, ProjectRef, ProjectId]],
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_spaces/_resolvers.py:18]: **nitpick**: you're probably aware of it, but `Workspace`, `Project`, and `ProjectRef` cases won't be needed here after `sdk.list_workflow_runs()` is simplified to accept just `Optional[WorkspaceId]` and `Optional[ProjectId]`

**Review by alexjuda — COMMENTED** [review]
Added comments to the remaining files. I know some of them might be already of date given our discussions on Slack, but I wanted to have a trace so we don't overlook them.

File: src/orquestra/sdk/_base/_spaces/_resolvers.py (hunk 2 above)
Comment (BenjaminMummery) [reply src/orquestra/sdk/_base/_spaces/_resolvers.py]: Get_space_ids is removed in [99587a9](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/153/commits/99587a9a288ab991de54a6414336d8ad42acd7d7) as we no longer need to worry about clashes.

File: src/orquestra/sdk/_base/cli/_dorq/_workflow/_list.py [hunk 3]
```diff
@@ -99,28 +99,49 @@ def _on_cmd_call_with_exceptions(
...
+                if resolved_project_id is not None:
+                    # We have a project, so we can specify the full workspace/project
+                    # via a ProjectRef
+                    project = ProjectRef(
+                        workspace_id=resolved_workspace_id,
+                        project_id=resolved_project_id,
+                    )
+                else:
+                    # We don't have a project so we'll pass the workspace
+                    workspace = resolved_workspace_id
```
[the remark this answers is not in the document]
Comment (BenjaminMummery) [reply src/orquestra/sdk/_base/cli/_dorq/_workflow/_list.py]: Good point, fixed in [99587a9](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/153/commits/99587a9a288ab991de54a6414336d8ad42acd7d7)

**Review by alexjuda — APPROVED** [review]

**Pull request merged**